File: src/command-risk.ts
import type { RiskLevel } from './types.js';

export interface CommandRiskResult {
  level: RiskLevel;
  reason: string;
}

export interface CommandRiskPattern {
  pattern: RegExp;
  risk: RiskLevel;
  reason: string;
}

export interface CommandRiskConfig {
  allowlist: Array<{ pattern: string; risk: RiskLevel; reason: string }>;
  blocklist: Array<{ pattern: string; risk: RiskLevel; reason: string }>;
}

const DANGER_PATTERNS: CommandRiskPattern[] = [
  { pattern: /\brm\s+(-[a-zA-Z]*r[a-zA-Z]*f|--force.*--recursive|--recursive.*--force|-[a-zA-Z]*f[a-zA-Z]*r)\b/, risk: 'danger', reason: 'Destructive recursive deletion' },
  { pattern: /\brm\s+-rf\b/, risk: 'danger', reason: 'Destructive recursive deletion' },
  { pattern: /\bsudo\b/, risk: 'danger', reason: 'Elevated privilege command' },
  { pattern: /\bcurl\b/, risk: 'danger', reason: 'Network request to external URL' },
  { pattern: /\bwget\b/, risk: 'danger', reason: 'Network download from external URL' },
  { pattern: /\bchmod\s+777\b/, risk: 'danger', reason: 'Setting world-writable permissions' },
  { pattern: /\bchmod\s+\+s\b/, risk: 'danger', reason: 'Setting setuid/setgid bit' },
  { pattern: /\bdd\s+/, risk: 'danger', reason: 'Low-level disk operation' },
  { pattern: /\bmkfs\b/, risk: 'danger', reason: 'Filesystem format operation' },
  { pattern: /\bkill\s+-9\b/, risk: 'danger', reason: 'Force-killing process' },
  { pattern: /\bkill\s+-SIGKILL\b/, risk: 'danger', reason: 'Force-killing process' },
  { pattern: /\bformat\b/, risk: 'danger', reason: 'Disk format operation' },
  { pattern: /\bdel\s+\/[fF]\s+\/[sS]\b/, risk: 'danger', reason: 'Destructive recursive deletion (Windows)' },
  { pattern: /\beval\b/, risk: 'danger', reason: 'Dynamic code execution' },
  { pattern: /\bnc\s+-l\b/, risk: 'danger', reason: 'Opening network listener' },
  { pattern: /\bssh\b.*@/, risk: 'danger', reason: 'Remote SSH connection' },
  { pattern: /\bscp\b/, risk: 'danger', reason: 'Remote file copy' },
  { pattern: /\bgit\s+push\s+--force\b/, risk: 'danger', reason: 'Force-pushing to remote repository' },
  { pattern: /\bgit\s+reset\s+--hard\b/, risk: 'danger', reason: 'Hard reset discards changes' },
];

const CAUTION_PATTERNS: CommandRiskPattern[] = [
  { pattern: /\bnpm\s+install\b/, risk: 'caution', reason: 'Installing npm package' },
  { pattern: /\bnpm\s+i\b/, risk: 'caution', reason: 'Installing npm package' },
  { pattern: /\bpip\s+install\b/, risk: 'caution', reason: 'Installing Python package' },
  { pattern: /\bbrew\s+install\b/, risk: 'caution', reason: 'Installing Homebrew package' },
  { pattern: /\bapt\s+install\b/, risk: 'caution', reason: 'Installing system package' },
  { pattern: /\bapt-get\s+install\b/, risk: 'caution', reason: 'Installing system package' },
  { pattern: /\byarn\s+add\b/, risk: 'caution', reason: 'Installing yarn package' },
  { pattern: /\bpnpm\s+add\b/, risk: 'caution', reason: 'Installing pnpm package' },
  { pattern: /\bgit\s+push\b/, risk: 'caution', reason: 'Pushing to remote repository' },
  { pattern: /\bdocker\s+run\b/, risk: 'caution', reason: 'Running Docker container' },
  { pattern: /\bdocker\s+exec\b/, risk: 'caution', reason: 'Executing in Docker container' },
  { pattern: /\bchmod\b/, risk: 'caution', reason: 'Changing file permissions' },
  { pattern: /\bchown\b/, risk: 'caution', reason: 'Changing file ownership' },
  { pattern: /\brm\b/, risk: 'caution', reason: 'Deleting files' },
  { pattern: /\bgit\s+checkout\s+--\b/, risk: 'caution', reason: 'Discarding file changes' },
  { pattern: /\bnpx\b/, risk: 'caution', reason: 'Executing remote npm package' },
];

const SAFE_PATTERNS: CommandRiskPattern[] = [
  { pattern: /\bgit\s+add\b/, risk: 'safe', reason: 'Staging files' },
  { pattern: /\bgit\s+status\b/, risk: 'safe', reason: 'Checking git status' },
  { pattern: /\bgit\s+diff\b/, risk: 'safe', reason: 'Viewing git diff' },
  { pattern: /\bgit\s+log\b/, risk: 'safe', reason: 'Viewing git log' },
  { pattern: /\bgit\s+branch\b/, risk: 'safe', reason: 'Managing branches' },
  { pattern: /\bgit\s+stash\b/, risk: 'safe', reason: 'Stashing changes' },
  { pattern: /\bls\b/, risk: 'safe', reason: 'Listing directory' },
  { pattern: /\bcat\b/, risk: 'safe', reason: 'Reading file' },
  { pattern: /\becho\b/, risk: 'safe', reason: 'Echoing output' },
  { pattern: /\bpwd\b/, risk: 'safe', reason: 'Printing working directory' },
  { pattern: /\bnode\b/, risk: 'safe', reason: 'Running Node.js' },
  { pattern: /\bnpm\s+test\b/, risk: 'safe', reason: 'Running tests' },
  { pattern: /\bnpm\s+run\s+build\b/, risk: 'safe', reason: 'Building project' },
  { pattern: /\bnpm\s+run\s+dev\b/, risk: 'safe', reason: 'Running dev server' },
  { pattern: /\bnpm\s+run\s+lint\b/, risk: 'safe', reason: 'Running linter' },
  { pattern: /\btsc\b/, risk: 'safe', reason: 'Running TypeScript compiler' },
  { pattern: /\bvitest\b/, risk: 'safe', reason: 'Running tests' },
  { pattern: /\bmkdir\b/, risk: 'safe', reason: 'Creating directory' },
  { pattern: /\bcp\b/, risk: 'safe', reason: 'Copying files' },
  { pattern: /\bmv\b/, risk: 'safe', reason: 'Moving files' },
  { pattern: /\bhead\b/, risk: 'safe', reason: 'Reading file head' },
  { pattern: /\btail\b/, risk: 'safe', reason: 'Reading file tail' },
  { pattern: /\bgrep\b/, risk: 'safe', reason: 'Searching content' },
  { pattern: /\bfind\b/, risk: 'safe', reason: 'Finding files' },
];

export function assessCommandRisk(
  executable: string,
  args: string[],
  config?: CommandRiskConfig,
): CommandRiskResult {
  const fullCommand = [executable, ...args].join(' ');

  // Check custom config patterns first (allowlist then blocklist)
  if (config) {
    for (const entry of config.allowlist) {
      if (new RegExp(entry.pattern).test(fullCommand)) {
        return { level: entry.risk, reason: entry.reason };
      }
    }
    for (const entry of config.blocklist) {
      if (new RegExp(entry.pattern).test(fullCommand)) {
        return { level: entry.risk, reason: entry.reason };
      }
    }
  }

  // Check danger patterns first (highest priority)
  for (const { pattern, reason } of DANGER_PATTERNS) {
    if (pattern.test(fullCommand)) {
      return { level: 'danger', reason };
    }
  }

  // Then safe patterns (before caution, to allow specific safe overrides)
  for (const { pattern, reason } of SAFE_PATTERNS) {
    if (pattern.test(fullCommand)) {
      return { level: 'safe', reason };
    }
  }

  // Then caution patterns
  for (const { pattern, reason } of CAUTION_PATTERNS) {
    if (pattern.test(fullCommand)) {
      return { level: 'caution', reason };
    }
  }

  // Unknown commands default to safe
  return { level: 'safe', reason: 'Standard command' };
}

File: unworldly/command_risk.py
"""Command risk assessment for Unworldly.

Classifies shell commands as safe, caution, or danger based on patterns.
Order: custom config > danger patterns > safe patterns > caution patterns.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional

from .types import RiskLevel


@dataclass
class CommandRiskResult:
    """Result of a command risk assessment."""
    level: RiskLevel
    reason: str


@dataclass
class CommandRiskPattern:
    """A pattern for matching commands to risk levels."""
    pattern: re.Pattern[str]
    risk: RiskLevel
    reason: str


@dataclass
class CommandRiskConfig:
    """Custom command risk configuration (allowlist/blocklist)."""
    allowlist: list[dict] = field(default_factory=list)
    blocklist: list[dict] = field(default_factory=list)


# --- DANGER patterns: destructive, privileged, network ---

DANGER_PATTERNS: list[CommandRiskPattern] = [
    CommandRiskPattern(re.compile(r"\brm\s+(-[a-zA-Z]*r[a-zA-Z]*f|--force.*--recursive|--recursive.*--force|-[a-zA-Z]*f[a-zA-Z]*r)\b"), RiskLevel.DANGER, "Destructive recursive deletion"),
    CommandRiskPattern(re.compile(r"\brm\s+-rf\b"), RiskLevel.DANGER, "Destructive recursive deletion"),
    CommandRiskPattern(re.compile(r"\bsudo\b"), RiskLevel.DANGER, "Elevated privilege command"),
    CommandRiskPattern(re.compile(r"\bcurl\b"), RiskLevel.DANGER, "Network request to external URL"),
    CommandRiskPattern(re.compile(r"\bwget\b"), RiskLevel.DANGER, "Network download from external URL"),
    CommandRiskPattern(re.compile(r"\bchmod\s+777\b"), RiskLevel.DANGER, "Setting world-writable permissions"),
    CommandRiskPattern(re.compile(r"\bchmod\s+\+s\b"), RiskLevel.DANGER, "Setting setuid/setgid bit"),
    CommandRiskPattern(re.compile(r"\bdd\s+"), RiskLevel.DANGER, "Low-level disk operation"),
    CommandRiskPattern(re.compile(r"\bmkfs\b"), RiskLevel.DANGER, "Filesystem format operation"),
    CommandRiskPattern(re.compile(r"\bkill\s+-9\b"), RiskLevel.DANGER, "Force-killing process"),
    CommandRiskPattern(re.compile(r"\bkill\s+-SIGKILL\b"), RiskLevel.DANGER, "Force-killing process"),
    CommandRiskPattern(re.compile(r"\bformat\b"), RiskLevel.DANGER, "Disk format operation"),
    CommandRiskPattern(re.compile(r"\bdel\s+/[fF]\s+/[sS]\b"), RiskLevel.DANGER, "Destructive recursive deletion (Windows)"),
    CommandRiskPattern(re.compile(r"\beval\b"), RiskLevel.DANGER, "Dynamic code execution"),
    CommandRiskPattern(re.compile(r"\bnc\s+-l\b"), RiskLevel.DANGER, "Opening network listener"),
    CommandRiskPattern(re.compile(r"\bssh\b.*@"), RiskLevel.DANGER, "Remote SSH connection"),
    CommandRiskPattern(re.compile(r"\bscp\b"), RiskLevel.DANGER, "Remote file copy"),
    CommandRiskPattern(re.compile(r"\bgit\s+push\s+--force\b"), RiskLevel.DANGER, "Force-pushing to remote repository"),
    CommandRiskPattern(re.compile(r"\bgit\s+reset\s+--hard\b"), RiskLevel.DANGER, "Hard reset discards changes"),
]

# --- CAUTION patterns: package installs, docker, permission changes ---

CAUTION_PATTERNS: list[CommandRiskPattern] = [
    CommandRiskPattern(re.compile(r"\bnpm\s+install\b"), RiskLevel.CAUTION, "Installing npm package"),
    CommandRiskPattern(re.compile(r"\bnpm\s+i\b"), RiskLevel.CAUTION, "Installing npm package"),
    CommandRiskPattern(re.compile(r"\bpip\s+install\b"), RiskLevel.CAUTION, "Installing Python package"),
    CommandRiskPattern(re.compile(r"\bbrew\s+install\b"), RiskLevel.CAUTION, "Installing Homebrew package"),
    CommandRiskPattern(re.compile(r"\bapt\s+install\b"), RiskLevel.CAUTION, "Installing system package"),
    CommandRiskPattern(re.compile(r"\bapt-get\s+install\b"), RiskLevel.CAUTION, "Installing system package"),
    CommandRiskPattern(re.compile(r"\byarn\s+add\b"), RiskLevel.CAUTION, "Installing yarn package"),
    CommandRiskPattern(re.compile(r"\bpnpm\s+add\b"), RiskLevel.CAUTION, "Installing pnpm package"),
    CommandRiskPattern(re.compile(r"\bgit\s+push\b"), RiskLevel.CAUTION, "Pushing to remote repository"),
    CommandRiskPattern(re.compile(r"\bdocker\s+run\b"), RiskLevel.CAUTION, "Running Docker container"),
    CommandRiskPattern(re.compile(r"\bdocker\s+exec\b"), RiskLevel.CAUTION, "Executing in Docker container"),
    CommandRiskPattern(re.compile(r"\bchmod\b"), RiskLevel.CAUTION, "Changing file permissions"),
    CommandRiskPattern(re.compile(r"\bchown\b"), RiskLevel.CAUTION, "Changing file ownership"),
    CommandRiskPattern(re.compile(r"\brm\b"), RiskLevel.CAUTION, "Deleting files"),
    CommandRiskPattern(re.compile(r"\bgit\s+checkout\s+--\b"), RiskLevel.CAUTION, "Discarding file changes"),
    CommandRiskPattern(re.compile(r"\bnpx\b"), RiskLevel.CAUTION, "Executing remote npm package"),
]

# --- SAFE patterns: read-only commands, standard dev tools ---

SAFE_PATTERNS: list[CommandRiskPattern] = [
    CommandRiskPattern(re.compile(r"\bgit\s+add\b"), RiskLevel.SAFE, "Staging files"),
    CommandRiskPattern(re.compile(r"\bgit\s+status\b"), RiskLevel.SAFE, "Checking git status"),
    CommandRiskPattern(re.compile(r"\bgit\s+diff\b"), RiskLevel.SAFE, "Viewing git diff"),
    CommandRiskPattern(re.compile(r"\bgit\s+log\b"), RiskLevel.SAFE, "Viewing git log"),
    CommandRiskPattern(re.compile(r"\bgit\s+branch\b"), RiskLevel.SAFE, "Managing branches"),
    CommandRiskPattern(re.compile(r"\bgit\s+stash\b"), RiskLevel.SAFE, "Stashing changes"),
    CommandRiskPattern(re.compile(r"\bls\b"), RiskLevel.SAFE, "Listing directory"),
    CommandRiskPattern(re.compile(r"\bcat\b"), RiskLevel.SAFE, "Reading file"),
    CommandRiskPattern(re.compile(r"\becho\b"), RiskLevel.SAFE, "Echoing output"),
    CommandRiskPattern(re.compile(r"\bpwd\b"), RiskLevel.SAFE, "Printing working directory"),
    CommandRiskPattern(re.compile(r"\bnode\b"), RiskLevel.SAFE, "Running Node.js"),
    CommandRiskPattern(re.compile(r"\bnpm\s+test\b"), RiskLevel.SAFE, "Running tests"),
    CommandRiskPattern(re.compile(r"\bnpm\s+run\s+build\b"), RiskLevel.SAFE, "Building project"),
    CommandRiskPattern(re.compile(r"\bnpm\s+run\s+dev\b"), RiskLevel.SAFE, "Running dev server"),
    CommandRiskPattern(re.compile(r"\bnpm\s+run\s+lint\b"), RiskLevel.SAFE, "Running linter"),
    CommandRiskPattern(re.compile(r"\btsc\b"), RiskLevel.SAFE, "Running TypeScript compiler"),
    CommandRiskPattern(re.compile(r"\bvitest\b"), RiskLevel.SAFE, "Running tests"),
    CommandRiskPattern(re.compile(r"\bmkdir\b"), RiskLevel.SAFE, "Creating directory"),
    CommandRiskPattern(re.compile(r"\bcp\b"), RiskLevel.SAFE, "Copying files"),
    CommandRiskPattern(re.compile(r"\bmv\b"), RiskLevel.SAFE, "Moving files"),
    CommandRiskPattern(re.compile(r"\bhead\b"), RiskLevel.SAFE, "Reading file head"),
    CommandRiskPattern(re.compile(r"\btail\b"), RiskLevel.SAFE, "Reading file tail"),
    CommandRiskPattern(re.compile(r"\bgrep\b"), RiskLevel.SAFE, "Searching content"),
    CommandRiskPattern(re.compile(r"\bfind\b"), RiskLevel.SAFE, "Finding files"),
]


def assess_command_risk(
    executable: str,
    args: list[str],
    config: Optional[CommandRiskConfig] = None,
) -> CommandRiskResult:
    """Assess the risk level of a shell command.

    Check order: custom config (allowlist then blocklist) > danger > safe > caution.
    Unknown commands default to safe.
    """
    full_command = " ".join([executable] + args)

    # Check custom config patterns first (allowlist then blocklist)
    if config is not None:
        for entry in config.allowlist:
            if re.search(entry["pattern"], full_command):
                return CommandRiskResult(
                    level=RiskLevel(entry["risk"]),
                    reason=entry["reason"],
                )
        for entry in config.blocklist:
            if re.search(entry["pattern"], full_command):
                return CommandRiskResult(
                    level=RiskLevel(entry["risk"]),
                    reason=entry["reason"],
                )

    # Check danger patterns first (highest priority)
    for crp in DANGER_PATTERNS:
        if crp.pattern.search(full_command):
            return CommandRiskResult(level=RiskLevel.DANGER, reason=crp.reason)

    # Then safe patterns (before caution, to allow specific safe overrides)
    for crp in SAFE_PATTERNS:
        if crp.pattern.search(full_command):
            return CommandRiskResult(level=RiskLevel.SAFE, reason=crp.reason)

    # Then caution patterns
    for crp in CAUTION_PATTERNS:
        if crp.pattern.search(full_command):
            return CommandRiskResult(level=RiskLevel.CAUTION, reason=crp.reason)

    # Unknown commands default to safe
    return CommandRiskResult(level=RiskLevel.SAFE, reason="Standard command")
